This might break below program I posted on the tracker. I am slightly surprised there are no CI failures with this change.

```python
from unittest.mock import call, ANY, Mock

class Foo:

    def __eq__(self, other):
        if not isinstance(other, Foo):
            return False
        return True

m = Mock()
obj = Foo()
m(obj, 1)
m.assert_has_calls([call(ANY, 1)])
```

Without patch

```
➜  cpython git:(master) ./python.exe ../backups/bpo37555_1.py
```

With patch

```
➜  cpython git:(master) git checkout pr_14700 Lib/unittest/mock.py
➜  cpython git:(master) ✗ ./python.exe ../backups/bpo37555_1.py
Traceback (most recent call last):
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/../backups/bpo37555_1.py", line 14, in <module>
    m.assert_has_calls([call(ANY, 1)])
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/unittest/mock.py", line 900, in assert_has_calls
    raise AssertionError(
AssertionError: Calls not found.
Expected: [call(<ANY>, 1)]
Actual: [call(<__main__.Foo object at 0x10b4a5280>, 1)].
```

--- a/Lib/unittest/mock.py
+++ b/Lib/unittest/mock.py
@@ -337,7 +337,7 @@ class _CallList(list):
 
         for i in range(0, len_self - len_value + 1):
             sub_list = self[i:i+len_value]
-            if value == sub_list:
+            if sub_list == value:
                 return True
         return False
 

--- a/Lib/unittest/test/testmock/testhelpers.py
+++ b/Lib/unittest/test/testmock/testhelpers.py
@@ -63,7 +63,19 @@ class AnyTest(unittest.TestCase):
             ]
             self.assertEqual(expected, mock.mock_calls)
             self.assertEqual(mock.mock_calls, expected)
+            mock.assert_has_calls(expected)
 
+    def test_assert_has_calls_with_any_and_spec_set(self):
+        # This is a regression test for bpo-37555
+        class Foo(object):
+            def __eq__(self, other): pass
+            def __ne__(self, other): pass
+
+        mock = Mock(spec_set=Foo)
+        expected = [call(ANY)]
+        mock(Foo())
+
+        mock.assert_has_calls(expected)
 
 
 class CallTest(unittest.TestCase):

--- a/Lib/unittest/test/testmock/testhelpers.py
+++ b/Lib/unittest/test/testmock/testhelpers.py
@@ -322,6 +322,16 @@ class CallTest(unittest.TestCase):
         kall = call(1).method(2)(3).foo.bar.baz(4)(5).__int__()
         self.assertEqual(kall.call_list(), mock.mock_calls)
 
+    def test_call_list_with_any_comparison_order(self):
+        class Foo:
+            def __eq__(self, other): pass
+            def __ne__(self, other): pass
+
+        mock = MagicMock()
+        mock(Foo())
+
+        self.assertEqual(call(ANY).call_list(), mock.mock_calls)
+        self.assertEqual(mock.mock_calls, call(ANY).call_list())
 
     def test_call_any(self):
         self.assertEqual(call, ANY)
@@ -331,6 +341,16 @@ class CallTest(unittest.TestCase):
         self.assertEqual(m.mock_calls, [ANY])
         self.assertEqual([ANY], m.mock_calls)
 
+    def test_call_any_comparison_order(self):
+        class Foo:
+            def __eq__(self, other): pass
+            def __ne__(self, other): pass
+
+        m = MagicMock()
+        m(Foo())
+
+        self.assertEqual(m.mock_calls[0], call(ANY))
+        self.assertEqual(call(ANY), m.mock_calls[0])
 
     def test_two_args_call(self):
         args = _Call(((1, 2), {'a': 3}), two=True)

